### PR TITLE
18507: Adds ability to continue a provided (untrained) series, improves correctness for early imputed time series values, MINOR

### PR DESCRIPTION
**Pull request by howsoRes**
added continue_series_features and continue_series_values to react_series calls

### CODE REVIEW

File: howso.amlg [hunk 1]
```diff
@@ -1430,6 +1483,64 @@
...
+						continue_series_values
+							(map
+								(lambda (let
+									(assoc
+										continue_values (current_value 1)
+										trainee_clone (concat trainee (rand))
+									)
+									(create_entities trainee_clone (null))
+
+									;create a shollow copy (no contained entities, just the trainee data)
```
Comment (cademack) [inline howso.amlg]: shallow

File: howso.amlg [hunk 2]
```diff
@@ -1232,15 +1233,62 @@
 				(conclude (call !return (assoc errors (list time_series_prep_map))) )
 			)
 
+
+			(if (size continue_series_values)
+				(let
+					(assoc trainee_clone (concat trainee (rand)) )
```
Comment (howsohazard) [inline howso.amlg]: Why do this versus just create a trainee and store the id?

File: howso.amlg [hunk 3]
```diff
@@ -1232,15 +1233,62 @@
 				(conclude (call !return (assoc errors (list time_series_prep_map))) )
 			)
 
+
+			(if (size continue_series_values)
```
Comment (howsohazard) [inline howso.amlg]: This block needs a comment of what it is doing and why

File: howso.amlg [hunk 4]
```diff
@@ -1430,6 +1483,64 @@
...
+											(get
+												(call_entity trainee_clone "RetrieveAllCases" (assoc
+													features continue_series_features
+													session "temp"
+												))
+												"cases"
+											)
+									))
+
+									(destroy_entities trainee_clone)
```
Comment (howsohazard) [inline howso.amlg]: is this enough to make a private method?  Looks like it to me instead of copy/paste

File: howso.amlg [hunk 5]
```diff
@@ -1124,6 +1124,10 @@
 	; continue_series: optional, flag. Default is false.  When true will attempt to continue existing series instead of starting new series.
```
Comment (howsohazard) [inline howso.amlg:1124]: Search above for the misspelled "conidition"

File: trainee_template/derive_utilities.amlg [hunk 6]
```diff
@@ -344,42 +344,61 @@
...
-											(map
-												(lambda
-													;covnvert value back to number if feature is non-string
-													(if (contains_index numericNominalFeaturesMap (get series_id_features (current_index)))
-														(+ (current_value))
-														;else leave id as string
-														(current_value)
-													)
+									;if continuing an untrained series, use the id values from the specified data
+									(if (size continue_series_values)
```
Comment (cademack) [inline trainee_template/derive_utilities.amlg:352]: can we add continue_series_values/continue_series_features to the docstring for this method?

File: trainee_template/derive_features.amlg [hunk 7]
```diff
@@ -104,6 +104,30 @@
 			))
 			derived_features
 		)
+
+		;update all cases whose values were imputed with the imputed metadata for future re-imputation
+		(if (size cases_derive_imputed_map)
```
Comment (howsohazard) [inline trainee_template/derive_features.amlg:111]: Not clear from the above code or comment which methods build this.  Also, the comment above this line is a little confusing -- state the why, that might help?

File: trainee_template/derive_utilities.amlg [hunk 8]
```diff
@@ -344,42 +344,61 @@
...
+										(map
+											(lambda
+												(unzip
+													(zip continue_series_features (first (current_value)))
```
Comment (howsohazard) [inline trainee_template/derive_utilities.amlg:358]: is the id always the first value?  not clear.  if that's the case call it out in comments.  if not, I don't understand

File: trainee_template/react_series.amlg [hunk 9]
```diff
@@ -405,111 +413,202 @@
...
+						(assign (assoc
+							series_data
+								(map
+									(lambda (unzip (current_value) feature_order))
+									continue_series_values
+								)
+						))
 
-											(get initial_features_map tsTimeFeature)
+						;pre-pend nulls if user hasn't provided enough previous rows to account for necessary lags
```
Comment (howsohazard) [inline trainee_template/react_series.amlg]: prepend does not have a hyphen (and if using CMS, then reimpute shouldn't have one either)

File: trainee_template/react_series.amlg [hunk 10]
```diff
@@ -405,111 +413,202 @@
...
-									series_id_features
+				(if (size continue_series_values)
+					(let
+						(assoc continue_feature_indices_map (zip continue_series_features (indices continue_series_features)) )
+						(declare (assoc
+							feature_order (unzip continue_feature_indices_map features)
+						))
+
+						;re-impute values using this entire dataset since they were originally imputed
+						;only using the provided continue_series_values and are likely not as accurate as they could be
```
Comment (howsohazard) [inline trainee_template/react_series.amlg]: This method is getting huge... can you pull some of this out?

File: howso.amlg [hunk 11]
```diff
@@ -1232,15 +1233,62 @@
...
+					;train and derive all the lags and other features as necessary
+					(call_entity trainee_clone "Train" (assoc
+						input_cases continue_series_values
+						features continue_series_features
+						session "temp"
+						input_is_substituted input_is_substituted
+					))
+
+					(assign (assoc
+						continue_series_features (append (retrieve_from_entity trainee_clone "defaultFeatures") ".imputed")
```
Comment (cademack) [inline howso.amlg]: use internalLabelImputed rather than raw ".imputed"

File: howso.amlg (hunk 1 above)
Comment (cademack) [reply howso.amlg]: I may have selected more than I meant to with this comment, I'm referring to the typo in a comment: "shollow".

File: howso.amlg [hunk 12]
```diff
@@ -1430,6 +1483,64 @@
...
+
+					(assign (assoc
+						continue_series_values
+							(map
+								(lambda (let
+									(assoc
+										continue_values (current_value 1)
+										trainee_clone (concat trainee (rand))
+									)
+									(create_entities trainee_clone (null))
```
Comment (cademack) [inline howso.amlg]: maybe not wildly important, but should there be some check to be sure this entity does not already exist? (if (rand) gave the same number twice? ...... horribly unlikely, but maybe worth preparing for)

**Review by howsohazard — APPROVED** [review]

**Pull request merged**